internal/http/services/appprovider/appprovider.go

```diff
@@ -0,0 +1,232 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ocmd
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	gateway "github.com/cs3org/go-cs3apis/cs3/gateway/v1beta1"
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/internal/http/services/ocmd"
+	"github.com/cs3org/reva/pkg/rgrpc/status"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/rhttp/global"
+	"github.com/cs3org/reva/pkg/sharedconf"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+func init() {
+	global.Register("appprovider", New)
+}
+
+// Config holds the config options that need to be passed down to all ocdav handlers
+type Config struct {
+	Prefix         string `mapstructure:"prefix"`
+	GatewaySvc     string `mapstructure:"gatewaysvc"`
+	AccessTokenTTL int    `mapstructure:"access_token_ttl"`
+}
+
+func (c *Config) init() {
+	if c.Prefix == "" {
+		c.Prefix = "api/v0/wopi/open"
+	}
+	if c.AccessTokenTTL == 0 {
+		c.AccessTokenTTL = 86400
+	}
+	c.GatewaySvc = sharedconf.GetGatewaySVC(c.GatewaySvc)
+}
+
+type svc struct {
+	conf *Config
+}
+
+// New returns a new ocmd object
+func New(m map[string]interface{}, log *zerolog.Logger) (global.Service, error) {
+
+	conf := &Config{}
+	if err := mapstructure.Decode(m, conf); err != nil {
+		return nil, err
+	}
+	conf.init()
+
+	s := &svc{
+		conf: conf,
+	}
+	return s, nil
+}
+
+// Close performs cleanup.
+func (s *svc) Close() error {
+	return nil
+}
+
+func (s *svc) Prefix() string {
+	return s.conf.Prefix
+}
+
+func (s *svc) Unprotected() []string {
+	return []string{}
+}
+
+func (s *svc) Handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			ocmd.WriteError(w, r, ocmd.APIErrorUnimplemented, "only GET requests are supported", errors.New("only GET requests are supported"))
+			return
+		}
+
+		s.handleWopiOpen(w, r)
+	})
+}
+
+// WopiResponse holds the various fields to be returned for a wopi open call
+type WopiResponse struct {
+	WopiClientURL  string `json:"wopiclienturl"`
```

Comment:
If this is supposed to be a generic appprovider responding via HTTP, then the response should be similar to the GRPC one, just the `AppURL`. Why do we want to expose the different parts of a WOPI-enabled application address (which is WOPI specific)?

Reply:
I also think that this part is too WOPI specific in order to be the http variant of the GRPC `OpenInAppRequest`. The input and output should match the GRPC counterpart.